README.md
# Overview #

Aurora DMX is designed to be simple, intuitive, and easy to use. This Android app is used to control lighting devices on DMX-512 via WiFi using the ArtNet protocol or sACN/ E1.31 via multicast.

https://play.google.com/store/apps/details?id=com.AuroraByteSoftware.AuroraDMX

# Description #

Features:
- Simple UI
- Selectable channel color
- Cues
- Rename Cues
- Cue fade times
- Channel to dimmer patch
- ArtNet
- sACN/ E1.31
- Save multiple projects
- sACN Unicast protocol
- View channel level as 255 steps
- Name Channels
- Set specific channel level/ step
- RGB Color selector
- ArtNet Universe
- Preset channel levels
- Share your projects 
- Next Cue button.
- Cue Sheet


The main screen was designed to give the light board operator a similar feel to what a production full size board would give them. It contains channels with channel number, level in percent, a level slider, and an edit button. Along the bottom is the cue list. Add Cue will create a light cue of the current channel levels and append it to the end of the cue list. If the use long presses on an existing cue they are able to edit it. Edit features are insert a new cue, remove cue, rename the cue, and change fade up and down times. In the settings menu ArtNet server is selected from a discovered list of nodes also manual entry is allowed. Default cue fade times can be assigned along with the color of the channel faders. Channel to dimmer patching is allowed in the patch view. One channel can be assigned to as many dimmers as you want. 

Save project will save current state of channels, patch, and cues to a user assigned name. Load project will open a previously saved project. A long press on the project name will prompt for delete. The current project is saved when exiting or switching project. The current project name is displayed at the top right of the main page.

The free version allows only 5 channels, an in-app purchase allows all 512 channels. The paid and free version allows patching to all 512 dimmers. 

If you any new features or different DMX protocols you would like to see just send me an email and I would love to get them in there.

Two ways to get signal to a DMX512 line:
Easiest: ENTTEC's ODE with a wireless router. 
Cheapest: Raspberry Pi running Open Lighting Architecture with ENTTEC's Open DMX USB and a wireless router.

Beta: https://play.google.com/apps/testing/com.AuroraByteSoftware.AuroraDMX
Donate: https://www.paypal.me/DanFredell


![Screenshot](/Pictures/3.0/Screenshot_20170216-192855.png)

# Development Setup #

* Download Android Studio 
* Open the AuroraDMX project

# Contributions #

* Create a Pull Request

# Thanks #

[![BrowserStack](/Pictures/Browserstack-logo.svg?raw=true&sanitize=true "BrowserStack")](https://www.browserstack.com/)

# License # 

GNU General Public License v3 (GPL-3) 



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [4.6.3-62]
### Fixed
- Fail to load previous projects

## [4.6.2-61]
### Fixed
- Crash on Reset settings

## [4.6.1-60]
### Fixed
- NPE on app close

## [4.6.0-58]
### Added
- Keep screen on to allow continual ArtNet sending
### Fixed
- Updated Android SDK to v30
- Use Gradle for ArtNetStack dependency management

## [4.5.0-57]
### Added
- Allow chase time less than a second
### Fixed
- When opening files
- Starting network before the app is fully loaded
- Fading cues

## [4.4.0-56]
### Added
- Allow cue fade time less than a second
### Fixed
- Crash when reordering Cues

## [4.3.0-55]
### Added
- Park Fixture. Toggle if the fixture value will be changed with cues and chases.

## [4.2.1-53]
### Fixed
- Added a default number of 0 to config popups when a blank value

## [4.2.0-52]
### Added
- GH-#12 Color chase buttons for easier finding
- GH-#11 Reorder chase and redesigned the popup reorder for cues
### Fixed
- Exiting chase crashes when you have more than a page of chase buttons

## [4.1.0-51]
### Added
- GH-#10 Add an import button in case your Android can't auto open the .AuroraDMX file
- When opening a project file, now it saves the project as the file name

## [4.0.5-50]
### Fixed
- Screen rotate looses channel levels

## [4.0.4-49]
### Fixed
- NPE on chase fade.
- Better chase fade handling with call backs.
- Prevent NPE on alCues.
- Look and feel black menu text fix.

## [4.0.3-48]
### Fixed
- Crash when deleting cues from IndexOutOfBounds
- Crash when loading ArtNet scan from IllegalState

## [4.0.2-47]
### Changed
- Updated icon to support different shapes

## [4.0.1-46]
### Fixed
- ConcurrentModificationException when viewing ArtNet Servers

### Changed
- How Billing is loaded in Settings so instant install can work
- Optimized imports

## [4.0.0-45]
### Added
- Chases that will chain together cues for a continuous fade or dramatic event.

### Changed
- Merged the cue fade up and down into one value

### Fixed
- SecurityException when opening a file
- Invalid manual ArtNET address
- NPE when sorting ArtNET servers
- More reliable UI during fade by using Handler instead of TimerTask

### Security
- Billing API updates
- Prevent file path traversal when sharing projects

## [3.1.1-39] 2017-06-17
### Fixed
- ArtNet scanning
- Modifying cues on Cue Sheet

## [3.1-38] 2017-05-24
### Added
- Cue Sheet

## [3.0-35] 2017-02-24
### Added
- Next Cue button.
- Allow Interrupting a fading Cue.
- Font Awesome icons

### Fixed
- Nexus 10 only showing one column

### Changed
- Redesign layout. 
- Use icons instead of names Update and Reorder existing Cues. 
- Use XML for UI 

## [2.8-30] 2016-11-27
### Added
- CID
- Improved sACN

## [2.7-28] 2016-10-24
### Fixed
- Fix 2 Console Errors java.lang.ArrayIndexOutOfBoundsException: length=512; index=512 	at com.AuroraByteSoftware.AuroraDMX.MainActivity.getCurrentDimmerLevels(MainActivity.java:352) 
 

## [2.3]
### Changed
- Raise max fixtures back to 512

## [2.2]
### Added
- ArtNet Port/Universe

## [2.1] 
### Added
- Fade time of 0.

### Changed 
- Improved fading performance.

## [2.0]
### Added
- RGB Fixtures

## [1.8-16]
### Added
- Edit button to channels
- Names to channels

### Remmoved
- Bump button

## [1.7-15]
### Fixed
- Bump button bug

## [1.7-14]
### Added
- View the brightness as 255 steps instead of just 100%

## [1.6-13]
### Added
- Network folder


app/src/main/java/com/AuroraByteSoftware/AuroraDMX/SACNData.java
package com.AuroraByteSoftware.AuroraDMX;

import java.io.UnsupportedEncodingException;
import java.lang.reflect.Method;

/**
 * 
 * @author Dan Fredell
 * April 26, 2014
 *
 */
class SACNData {

	private byte universe = 1;
	private byte sequenceNumber = 0x00;
	private final byte[] message;

	/**
	 * Initate the aSCN packet. Use message via pass by reference. 
	 * @param a_message the universe
	 * @param a_universe the data
	 * @throws UnsupportedEncodingException
	 */
	public SACNData(byte[] a_message, int a_universe)
			throws UnsupportedEncodingException {
		universe = (byte) a_universe;
		message = a_message;
		addRootLayer();
		addFrameLayer();
		addDMPLayer();
	}

	void addRootLayer() {
		// Preamble Size
		message[0] = (byte) 0x00;
		message[1] = (byte) 0x10;
		// Post-amble Size
		message[2] = (byte) 0x00;
		message[3] = (byte) 0x00;
		// ACN Packet Identifier
		message[4] = (byte) 0x41;
		message[5] = (byte) 0x53;
		message[6] = (byte) 0x43;
		message[7] = (byte) 0x2d;
		message[8] = (byte) 0x45;
		message[9] = (byte) 0x31;
		message[10] = (byte) 0x2e;
		message[11] = (byte) 0x31;
		message[12] = (byte) 0x37;
		message[13] = (byte) 0x00;
		message[14] = (byte) 0x00;
		message[15] = (byte) 0x00;
		// Flags and Length
		message[16] = ((byte) 0x72);
		message[17] = ((byte) 0x6e);
		// Vector
		message[18] = ((byte) 0x00);
		message[19] = ((byte) 0x00);
		message[20] = ((byte) 0x00);
		message[21] = ((byte) 0x04);
		// CID 22-37
		//TODO generate this from the serial number
		try {
			Class<?> c = Class.forName("android.os.SystemProperties");
			Method get = c.getMethod("get", String.class);
			String serial = (String) get.invoke(c, "ro.serialno");
			serial = serial.length() > 16 ? serial.substring(0, 16) : serial;
			byte[] serialBytes = serial.getBytes("UTF-8");
			System.arraycopy(serialBytes, 0, message, 22, serialBytes.length);
		} catch (Throwable e) {
			// TODO Auto-generated catch block
		}
	}

	void addFrameLayer() throws UnsupportedEncodingException {
		// Flags and Length
		message[38] = (byte) 0x72;
		message[39] = (byte) 0x58;
		// Vector
		message[40] = (byte) 0x00;
		message[41] = (byte) 0x00;
		message[42] = (byte) 0x00;
		message[43] = (byte) 0x02;
		// Source Name 44-107
		String name = "AuroraDMX";
		byte[] nameBytes = name.getBytes("UTF-8");
		System.arraycopy(nameBytes, 0, message, 44, nameBytes.length);
		// Priority
		message[108] = (byte) 100;
		// Reserved
		message[109] = (byte) 0x00;
		message[110] = (byte) 0x00;
		// Sequence Number
		//message[111] = sequenceNumber++;
		// Options
		message[112] = (byte) 0x00;
		// Universe
		message[113] = (byte) 0x00;
		message[114] = universe;

	}

	void addDMPLayer() {
		// Flags and Length
		message[115] = (byte) 0x72;
		message[116] = (byte) 0x0b;
		// Vector
		message[117] = (byte) 0x02;
		// Address Type & Data Type
		message[118] = (byte) 0xa1;
		// First Property Address
		message[119] = (byte) 0x00;
		message[120] = (byte) 0x00;
		// Address messagecrement
		message[121] = (byte) 0x00;
		message[122] = (byte) 0x01;
		// Address messagecrement
		message[123] = (byte) 0x02;
		message[124] = (byte) 0x01;
	}

	/**
	 * DMX data in int array level range 0-255, array range 0-511
	 * Array to channels are off by one
	 * @param levels 0-255 on each
	 */
	public void addDMXData(int levels[]) {
		// Property values. START Code + Data
		for (int x = 0; x < 512 && x < levels.length; x++) {
			message[126 + x] = (byte) levels[x];
		}
		// Sequence Number
		message[111] = sequenceNumber++;
	}

}
